### PR TITLE
AphasiaBank ASR recipe

### PR DESCRIPTION
**Pull request by tjysdsg**
English ASR recipe for [AphasiaBank](https://aphasia.talkbank.org/)

Current SOTA: https://aphasia.talkbank.org/publications/2021/Torre21.pdf

TODO:
- [x] E-Branchformer
- [x] WavLM
- [x] Upload to HuggingFace

### CODE REVIEW

File: egs2/TEMPLATE/asr1/db.sh [hunk 1]
```diff
@@ -10,6 +10,7 @@ AISHELL4=downloads
 ALFFA=downloads
 AMERICASNLP22=downloads
 AN4=downloads
+APHASIABANK=/ocean/projects/cis210027p/shared/corpora/AphasiaBank/English
```
Comment (tjysdsg) [inline egs2/TEMPLATE/asr1/db.sh:13]: TODO: replace absolute path

File: egs2/aphasiabank/asr1/cmd.sh [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+# If jobs failed, your configuration might be wrong for your environment.
+#
+#
+# The official documentation for run.pl, queue.pl, slurm.pl, and ssh.pl:
+#   "Parallelization in Kaldi": http://kaldi-asr.org/doc/queue.html
+# =========================================================~
+
+
+# Select the backend used by run.sh from "local", "stdout", "sge", "slurm", or "ssh"
+cmd_backend='slurm'
```
Comment (tjysdsg) [inline egs2/aphasiabank/asr1/cmd.sh:31]: TODO: revert changes

File: egs2/aphasiabank/asr1/conf/slurm.conf [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+default time=8:00:00
+option time=* --time $0
+option mem=* --mem-per-cpu $0
+option mem=0
+option num_threads=* --cpus-per-task $0
+option num_threads=1 --cpus-per-task 1
+option num_nodes=* --nodes $0
+default gpu=0
+option gpu=0 -p RM-shared
+option gpu=* -p GPU-shared --gres=gpu:v100-32:$0 -c $0  # Recommend allocating more CPU than, or equal to the number of GPU
```
Comment (tjysdsg) [inline egs2/aphasiabank/asr1/conf/slurm.conf:13]: TODO: revert changes

File: egs2/aphasiabank/asr1/run.sh [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+test_sets="test"
+include_control=false
+include_aphasia_type=true
+
+asr_tag=""
+if "${include_aphasia_type}"; then
+  asr_tag="train_asr_conformer_with_aphasia_type"
+fi
+
+# asr_config=conf/tuning/train_asr_transformer.yaml
```
Comment (tjysdsg) [inline egs2/aphasiabank/asr1/run.sh:19]: TODO: use symlink

File: egs2/TEMPLATE/asr1/db.sh (hunk 1 above)
Comment (simpleoier) [reply egs2/TEMPLATE/asr1/db.sh:13]: Maybe you should note somewhere (e.g. data.sh or README in the recipe) that this is only concerned with English.

File: egs2/aphasiabank/asr1/conf/decode.yaml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+beam_size: 10
+ctc_weight: 0.3
+lm_weight: 0.3
```
Comment (simpleoier) [inline egs2/aphasiabank/asr1/conf/decode.yaml:3]: Does LM help?

File: egs2/aphasiabank/asr1/run.sh [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+feats_normalize=global_mvn
+# feats_normalize=utt_mvn # if using HuBERT, https://github.com/espnet/espnet/issues/4006#issuecomment-1047898558
+
+# speed perturbation related
+# (train_set will be "${train_set}_sp" if speed_perturb_factors is specified)
+speed_perturb_factors="0.9 1.0 1.1"
+
+./asr.sh                                                                                                      \
+    --lang en                                                                                                 \
```
Comment (simpleoier) [inline egs2/aphasiabank/asr1/run.sh:35]: The position of "\" is a bit weird from this view, as it is too far from the end of the line. Please double-check if the space symbol is correctly used.

File: egs2/aphasiabank/asr1/local/data.sh [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+tmp=data/local
+mkdir -p $tmp
+
+# things that were manually done:
+# - Converting all *.mp4 and *.mp3 data into .wav files in their original directories
+# - Move all transcripts (including control group data) into ${APHASIABANK}/transcripts
+# - Speakers in train, test, and val set is hardcoded in split_train_test_val.py
+
+if [ ${stage} -le 1 ] && [ ${stop_stage} -ge 1 ]; then
+  # generate data/spk_info.txt
```
Comment (simpleoier) [inline egs2/aphasiabank/asr1/local/data.sh:54]: You may add some log for each stage.

**Review by ftshijt — COMMENTED** [review]
Could you also add an entry at `egs2/README.md` for the corpus?

File: egs2/aphasiabank/asr1/local/data.sh [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+if [ -z "${APHASIABANK}" ]; then
+  log "Error: \$APHASIABANK is not set in db.sh."
+  exit 2
+fi
+
+tmp=data/local
+mkdir -p $tmp
+
+# things that were manually done:
```
Comment (tjysdsg) [inline egs2/aphasiabank/asr1/local/data.sh:48]: Automate these

**Comment by sw005320** [comment]
Cool!
Again, E-branchformer outperformed conformer (@pyf98)

**Comment by pyf98** [comment]
Cool! Thanks for sharing the result.

**Comment by sw005320** [comment]
Amazing results!
Very cool!
Once it is ready, please make this PR from the draft status to the normal PR.
I'll check and merge this PR.

**Comment by tjysdsg** [comment]
> Amazing results! Very cool! Once it is ready, please make this PR from the draft status to the normal PR. I'll check and merge this PR.

Thanks!
I'm trying to pass all CI checks, once it's done I'll mark the PR as ready and request code review.

File: egs2/aphasiabank/asr1/conf/tuning/train_asr_ebranchformer_small_wavlm_large.yaml [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+encoder_conf:
+  output_size: 256
+  attention_heads: 4
+  linear_units: 1024
+  num_blocks: 12
+  dropout_rate: 0.1
+  positional_dropout_rate: 0.1
+  attention_dropout_rate: 0.1
+  layer_drop_rate: 0.1
+  input_layer: conv2d2 # subsampling rate = 2
```
[the remark this answers is not in the document]
Comment (tjysdsg) [reply egs2/aphasiabank/asr1/conf/tuning/train_asr_ebranchformer_small_wavlm_large.yaml:37]: Let me try not using downsampling just to be safe

**Comment by tjysdsg** [comment]
Just a minor comment, `Wavlm frontend` already has frame shift of 20ms. If you use conv2d2 here, it will be further downsampled with frame shift of 40ms. So if you want, you may try not to use downsampling in the encoder.

_Originally posted by @simpleoier in https://github.com/espnet/espnet/pull/4860#discussion_r1082707743_
            

@sw005320 @ftshijt Xuankai @simpleoier suggested that I write a `conv2d1` module with sampling rate = 1 and train again. I think I can do it in another PR, so this PR is now ready to be merged

**Comment by sw005320** [comment]
Thanks!

**Pull request merged**